release-24.3: sql: use memo metadata to add routines and UDTs to statement bundles

DrewKimball opened the pull request:
Backport 3/3 commits from #132147.
Backport 1/1 commits from #139193.

/cc @cockroachdb/release

---

#### opt: replace "udf" with "routine" in the metadata

This commit changes the naming of some methods and data structures used
in query plan metadata to refer to "routines" instead of "udfs". This
clarifies that stored procedures are also valid targets for metadata
tracking.

Epic: None

Release note: None

#### sql: use memo metadata to add routines to statement bundles

This commit updates the statement bundle logic to take advantage of the
information stored in the query plan metadata so that only relevant routines
are shown in `schema.sql` for a statement bundle. In addition, stored
procedures are now tracked in the metadata in addition to UDFs. This has
no impact on query-plan caching, since we currently do not cache plans
that invoke stored procedures.

Fixes #132142
Fixes #104976

Release note (bug fix): Fixed a bug that prevented the create statement
for a routine from being shown in a statement bundle. This happened when
the routine was created on a schema other than `public`. The bug has
existed since v23.1.

#### sql: use memo metadata to add UDTs to statement bundle

This commit modifies the statement bundle collection logic to use the
query plan metadata to determine which types to display in `schema.sql`.
This ensures that only types which are referenced by the query are shown.

Informs #104976

Release note: None

---

Release justification: statement bundles improvements

DrewKimball committed:
opt: replace "udf" with "routine" in the metadata

This commit changes the naming of some methods and data structures used
in query plan metadata to refer to "routines" instead of "udfs". This
clarifies that stored procedures are also valid targets for metadata
tracking.

Epic: None

Release note: None

DrewKimball committed:
sql: use memo metadata to add routines to statement bundles

This commit updates the statement bundle logic to take advantage of the
information stored in the query plan metadata so that only relevant routines
are shown in `schema.sql` for a statement bundle. In addition, stored
procedures are now tracked in the metadata in addition to UDFs. This has
no impact on query-plan caching, since we currently do not cache plans
that invoke stored procedures.

Fixes #132142
Fixes #104976

Release note (bug fix): Fixed a bug that prevented the create statement
for a routine from being shown in a statement bundle. This happened when
the routine was created on a schema other than `public`. The bug has
existed since v23.1.

DrewKimball committed:
sql: use memo metadata to add UDTs to statement bundle

This commit modifies the statement bundle collection logic to use the
query plan metadata to determine which types to display in `schema.sql`.
This ensures that only types which are referenced by the query are shown.

Informs #104976

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/136124)
<!-- Reviewable:end -->

mgartner reviewed (commented):
I think there's a couple things you might want to fixup on master before merging these backports.

> Reviewed 4 of 4 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-:-OCdY6JL0LWQ_lRJW0pF:ba63sc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/explain_bundle_test.go` line 430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdXnem2wTz5De5XYsl:-OCdXnem2wTz5De5XYsm:b-xcuf6q) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L430)):*
> ```Go
> 			t, fmt.Sprint(rows), "s.foo", func(name, contents string) error {
> 				if name == "schema.sql" {
> 					reg := regexp.MustCompile("s.foo")
> ```

`.` is a wild card match, so you probably want `"s\.foo"` instead
___
*[`pkg/sql/explain_bundle_test.go` line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdWqr61V6k5PzkKD-w:-OCdWqr61V6k5PzkKD-x:bvs1wxj) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L434)):*
> ```Go
> 						return errors.Errorf("could not find definition for 's.foo' function in schema.sql")
> 					}
> 					reg = regexp.MustCompile("^foo")
> ```

This will only match "foo" at the start of the line, not "CREATE FUNCTION foo ...", right? And the `foo` will be fully qualified in the bundle, so maybe you want something like `"public\.foo"`?
___
*[`pkg/sql/explain_bundle_test.go` line 465 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdX7UD281AhFBLP9mm:-OCdX7UD281AhFBLP9mn:b-lxwe1y) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L465)):*
> ```Go
> 						return errors.Errorf("could not find definition for 's.bar' procedure in schema.sql")
> 					}
> 					reg = regexp.MustCompile("^bar")
> ```

see my comment above
___
*[`pkg/sql/explain_bundle_test.go` line 469 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdXu3A7zAXsBARpkk0:-OCdXu3A7zAXsBARpkk1:b-20pe31) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bd2ea18ca1079b615f122e752a7cd536b0ab87/pkg/sql/explain_bundle_test.go#L469)):*
> ```Go
> 			t, fmt.Sprint(rows), "s.bar", func(name, contents string) error {
> 				if name == "schema.sql" {
> 					reg := regexp.MustCompile("s.bar")
> ```

`"s\.bar"`
___
*[`pkg/sql/explain_bundle.go` line 630 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdVa5A-iYqcQEDoV7N:-OCdVa5A-iYqcQEDoV7O:b-7oyntb) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle.go#L630)):*
> ```Go
> 	if mem.Metadata().HasUserDefinedRoutines() {
> 		// Get all relevant user-defined routines.
> 		var ids intsets.Fast
> ```

nit: `ids` doesn't seem necessary - why can't the code in the `ids.ForEach` anonymous function o in the `ForEachUserDefinedRoutine` anonymous function?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Friendly ping on pushing these backports over the line - I have some changes in-flight to improve MR stmt bundle recreation (which I'd like to backport) that depend on this.

> Reviewed 2 of 4 files at r1, 1 of 3 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-:-OERGxXL7VDcBXEcBAae:b-x5ivfz)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/explain_bundle.go` line 1015 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OERGduG6c43HUzTi_p-:-OERGduG6c43HUzTi_p0:bfdcg77) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bd2ea18ca1079b615f122e752a7cd536b0ab87/pkg/sql/explain_bundle.go#L1015)):*
> ```Go
> func (c *stmtEnvCollector) PrintCreateUDT(w io.Writer, id oid.Oid, redactValues bool) error {
> 	descID := catid.UserDefinedOIDToID(id)
> 	query := fmt.Sprintf("SELECT create_statement FROM crdb_internal.create_type_statements WHERE descriptor_id = %d", descID)
> ```

It might be nice to also pick up 644c458333e05bbb821dab1b9bb693d8003ef4fa into this backport.
___
*[`pkg/sql/explain_bundle.go` line 1024 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OERGT5BBPp4b1SfjJKC:-OERGT5BBPp4b1SfjJKD:bs5pjnz) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bd2ea18ca1079b615f122e752a7cd536b0ab87/pkg/sql/explain_bundle.go#L1024)):*
> ```Go
> 	}
> 	for _, cs := range createStatement {
> 		fmt.Fprintf(w, "%s\n", cs)
> ```

Noticed this while working on the MR stmt bundle recreations - we need to add `;` here. `SHOW CREATE ALL TYPES` was doing that for us, but `create_statement` doesn't.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[`pkg/sql/explain_bundle.go` line 1015 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OERGduG6c43HUzTi_p-:-OG2kVKe8ucUoVqgZ7Vb:bu8pdb5) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bd2ea18ca1079b615f122e752a7cd536b0ab87/pkg/sql/explain_bundle.go#L1015)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

It might be nice to also pick up 644c458333e05bbb821dab1b9bb693d8003ef4fa into this backport.
</blockquote></details>

Nevermind, we ended up reverting that change.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: minor fixup of a recent change

This commit addresses comments identified during the review of
a backport of 3fb9f918ab48e53b2feb6ec9c31b505272be30b1.

Release note: None

yuzefovich reviewed (commented):
I addressed all the comments on this PR in #139193.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-:-OGgcqsu5kQZS8lg4HMu:bnbldai)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/explain_bundle.go` line 630 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdVa5A-iYqcQEDoV7N:-OGg_KTB-EfKqaiUCMQj:bcabtj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle.go#L630)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

nit: `ids` doesn't seem necessary - why can't the code in the `ids.ForEach` anonymous function o in the `ForEachUserDefinedRoutine` anonymous function?
</blockquote></details>

The only benefit I see is that using `ids.ForEach` would give us deterministic order (in increasing Oids) of adding routines to the bundle, and assuming that only newer (with larger Oids) routines could use older ones (with smaller Oids), we need this for recreation. Still it would be possible to construct a schema such that routine with smaller Oid would depend on the one with larger Oid (via `CREATE OR REPLACE ...`).
___
*[`pkg/sql/explain_bundle_test.go` line 430 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdXnem2wTz5De5XYsl:-OGgbKhGC8gDd1RIBHHM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L430)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`.` is a wild card match, so you probably want `"s\.foo"` instead
</blockquote></details>

Done.
___
*[`pkg/sql/explain_bundle_test.go` line 434 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdWqr61V6k5PzkKD-w:-OGgboocE1taX-yHCgJo:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L434)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

This will only match "foo" at the start of the line, not "CREATE FUNCTION foo ...", right? And the `foo` will be fully qualified in the bundle, so maybe you want something like `"public\.foo"`?
</blockquote></details>

Good point, done.
___
*[`pkg/sql/explain_bundle_test.go` line 465 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdX7UD281AhFBLP9mm:-OGgc6xK7514Oiy_2Ixy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/77781b60c29aa1f39ad907a5add39e28d2547bd9/pkg/sql/explain_bundle_test.go#L465)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

see my comment above
</blockquote></details>

Done.
___
*[`pkg/sql/explain_bundle_test.go` line 469 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-OCdXu3A7zAXsBARpkk0:-OGgc6K79G8dRx8ESVOB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bd2ea18ca1079b615f122e752a7cd536b0ab87/pkg/sql/explain_bundle_test.go#L469)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

`"s\.bar"`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/136124#-:-OGpiUNw-VOj6770CjNj:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->
